src/main/java/com/yahoo/memory/BaseWritableMemoryImpl.java

```diff
@@ -0,0 +1,416 @@
+
+  //OTHER PRIMITIVE READ METHODS: compareTo, copyTo XXX
+  @Override
+  public int compareTo(final long thisOffsetBytes, final long thisLengthBytes, final Memory thatMem,
+      final long thatOffsetBytes, final long thatLengthBytes) {
+    state.checkValid();
+    checkBounds(thisOffsetBytes, thisLengthBytes, capacity);
+    final BaseWritableMemoryImpl that = (BaseWritableMemoryImpl) thatMem;
+    that.state.checkValid();
+    checkBounds(thatOffsetBytes, thatLengthBytes, that.capacity);
+    final long thisAdd = getCumulativeOffset(thisOffsetBytes);
+    final long thatAdd = that.getCumulativeOffset(thatOffsetBytes);
+    final Object thisObj = this.unsafeObj;
+    final Object thatObj = that.unsafeObj;
+    if (thisObj != thatObj || thisAdd != thatAdd) {
+      final long lenBytes = Math.min(thisLengthBytes, thatLengthBytes);
+      for (long i = 0; i < lenBytes; i++) {
+        final int thisByte = unsafe.getByte(thisObj, thisAdd + i);
+        final int thatByte = unsafe.getByte(thatObj, thatAdd + i);
+        if (thisByte < thatByte) {
+          return -1;
+        }
+        if (thisByte > thatByte) {
+          return 1;
+        }
+      }
+    }
+    return Long.compare(thisLengthBytes, thatLengthBytes);
+  }
+
+  @Override
+  public void copyTo(final long srcOffsetBytes, final WritableMemory destination,
+      final long dstOffsetBytes, final long lengthBytes) {
+    final BaseWritableMemoryImpl dst = (BaseWritableMemoryImpl) destination;
+    state.checkValid();
+    checkBounds(srcOffsetBytes, lengthBytes, capacity);
+    dst.state.checkValid();
+    checkBounds(dstOffsetBytes, lengthBytes, dst.capacity);
+    final long srcAdd = getCumulativeOffset(srcOffsetBytes);
+    final long dstAdd = destination.getCumulativeOffset(dstOffsetBytes);
+    copyMemory(unsafeObj, srcAdd, dst.unsafeObj, dstAdd, lengthBytes);
+  }
+
+  static void copyMemory(final Object srcUnsafeObj, final long srcAdd,
+      final Object dstUnsafeObj, final long dstAdd, final long lengthBytes) {
+    if (srcUnsafeObj != dstUnsafeObj) {
+      copyNonOverlappingMemory(srcUnsafeObj, srcAdd, dstUnsafeObj, dstAdd, lengthBytes);
+    } else {
+      copyMemorySlowPath(srcUnsafeObj, srcAdd, dstUnsafeObj, dstAdd, lengthBytes);
+    }
+  }
+
+  private static void copyMemorySlowPath(final Object srcUnsafeObj, final long srcAdd,
+      final Object dstUnsafeObj, final long dstAdd, final long lengthBytes) {
+    if (((srcAdd + lengthBytes) <= dstAdd) || ((dstAdd + lengthBytes) <= srcAdd)) {
+      copyNonOverlappingMemory(srcUnsafeObj, srcAdd, dstUnsafeObj, dstAdd, lengthBytes);
+      return;
+    }
+    if (srcAdd == dstAdd) {
+      throw new IllegalArgumentException(
+          "Attempt to copy a block of memory exactly in-place, should be a bug");
+    }
+    // If regions do overlap, fall back to unsafe.copyMemory, tolerating potentially long
+    // Time to Safe Point pauses.
+    unsafe.copyMemory(srcUnsafeObj, srcAdd, dstUnsafeObj, dstAdd, lengthBytes);
+  }
+
```

Comment:
This `copyMemoryCheckingNonOverlapping(...)` is only used for `put/get<primitive>Array(...)`.  
I don't understand the need.  The only way an overlap could occur would be if someone tried to use one of these methods to get or put an array into the same heap array that they wrapped.  I.e.

```java
    byte[] dst = new byte[128];
    WritableMemory wmem = WritableMemory.wrap(dst);
    wmem.getByteArray(0, dst, 64, 64);  //non-overlapping
    wmem.getByteArray(0, dst, 32, 64); //overlapping
```
This makes no sense to me, the user should use `copyTo()`!
We should check that the dst or src array is not the same as _this_, otherwise throw an error. 
Much simpler.

Reply:
Ok.

Reply:
Can you fix?

Reply:
Fixed.